src/consensus/validation.h

```diff
@@ -71,31 +71,32 @@ enum class BlockValidationResult {
 
 
 
-/** Base class for capturing information about block/transaction validation. This is subclassed
+/** Template for capturing information about block/transaction validation. This is instantiated
  *  by TxValidationState and BlockValidationState for validation information on transactions
  *  and blocks respectively. */
+template <typename Result>
 class ValidationState {
 private:
     enum mode_state {
         MODE_VALID,   //!< everything ok
         MODE_INVALID, //!< network rule violation (DoS value may be set)
         MODE_ERROR,   //!< run-time error
     } m_mode;
+    Result m_result;
     std::string m_reject_reason;
     std::string m_debug_message;
-protected:
-    void Invalid(const std::string &reject_reason="",
+public:
+    ValidationState() : m_mode(MODE_VALID) {}
```

Comment:
nit, in L85 do `} m_mode{MODE_VALID};` and drop constructor?

Reply:
The diff here is a bit misleading as I'm actually moving `Invalid` to the `public` section and not modifying the constructor. Is the convention to put this in a separate a commit? Or would it be considered outside the scope of the PR?

Reply:
Right, but the constructor could still be dropped.

Reply:
Understood. But that doesn't answer either of my questions... 🤔

Reply:
IMHO could be added in "Templatize ValidationState instead of subclassing", but no big deal.

Reply:
Ended up doing this separately in 7470037 as it gives cleaner diffs.